File: backend/app/services/page_service.py
from sqlalchemy import or_
from sqlalchemy.orm import Session

from app.models.page import Page
from app.schemas.page import (
    PageCreate,
    PageUpdate,
)
from app.utils.slug import slugify


def create_page(
    db: Session,
    page: PageCreate,
):
    page_data = page.model_dump()

    base_slug = slugify(page.title)
    slug = base_slug
    counter = 1

    while db.query(Page).filter(Page.slug == slug).first():
        slug = f"{base_slug}-{counter}"
        counter += 1

    page_data["slug"] = slug

    new_page = Page(**page_data)

    db.add(new_page)
    db.commit()
    db.refresh(new_page)

    return new_page


def get_all_pages(
    db: Session,
    skip: int = 0,
    limit: int = 10,
    search: str | None = None,
):
    query = db.query(Page)

    if search:
        query = query.filter(
            or_(
                Page.title.ilike(f"%{search}%"),
                Page.summary.ilike(f"%{search}%"),
                Page.content.ilike(f"%{search}%"),
            )
        )

    return (
        query
        .order_by(Page.id.desc())
        .offset(skip)
        .limit(limit)
        .all()
    )


def get_page_by_id(
    db: Session,
    page_id: int,
):
    return (
        db.query(Page)
        .filter(Page.id == page_id)
        .first()
    )


def get_page_by_slug(
    db: Session,
    slug: str,
):
    return (
        db.query(Page)
        .filter(Page.slug == slug)
        .first()
    )


def update_page(
    db: Session,
    page_id: int,
    page: PageUpdate,
):
    existing_page = get_page_by_id(
        db,
        page_id,
    )

    if not existing_page:
        return None

    page_data = page.model_dump()

    if existing_page.title != page.title:
        base_slug = slugify(page.title)
        slug = base_slug
        counter = 1

        while (
            db.query(Page)
            .filter(
                Page.slug == slug,
                Page.id != page_id,
            )
            .first()
        ):
            slug = f"{base_slug}-{counter}"
            counter += 1

        page_data["slug"] = slug

    for key, value in page_data.items():
        setattr(existing_page, key, value)

    db.commit()
    db.refresh(existing_page)

    return existing_page


def delete_page(
    db: Session,
    page_id: int,
):
    existing_page = get_page_by_id(
        db,
        page_id,
    )

    if not existing_page:
        return None

    db.delete(existing_page)
    db.commit()

    return existing_page

File: admin/src/services/pageService.ts
import api from "./api";

export interface PageData {
  id: number;
  title: string;
  slug: string;
  summary: string;
  content: string;
  cover_image: string | null;
  seo_title: string | null;
  seo_description: string | null;
  is_published: boolean;
  created_at: string;
  updated_at: string;
}

export interface PageUpdateData {
  title: string;
  summary: string;
  content: string;
  cover_image?: string | null;
  seo_title?: string | null;
  seo_description?: string | null;
  is_published: boolean;
}

export const getAllPages = async (): Promise<PageData[]> => {
  const response = await api.get("/pages/");
  return response.data;
};

export const getPageBySlug = async (
  slug: string
): Promise<PageData> => {
  const response = await api.get(`/pages/slug/${slug}`);
  return response.data;
};

export const getPageById = async (
  id: number
): Promise<PageData> => {
  const response = await api.get(`/pages/${id}`);
  return response.data;
};

export const updatePage = async (
  id: number,
  data: PageUpdateData
): Promise<PageData> => {
  const response = await api.put(
    `/pages/${id}`,
    data
  );

  return response.data;
};

export const createPage = async (
  data: PageUpdateData
): Promise<PageData> => {
  const response = await api.post(
    "/pages/",
    data
  );

  return response.data;
};

export const deletePage = async (
  id: number
): Promise<void> => {
  await api.delete(`/pages/${id}`);
};